File: nikkake_flutter/lib/screens/settings_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../constants/colors.dart';
import '../models/models.dart';
import '../providers/app_state.dart';
import '../providers/auth_controller.dart';
import '../widgets/ui.dart';
import 'login_screen.dart';

/// 設定。
///
/// この画面の主役は「バックアップするかどうか」の説明。
/// サインインは機能を開放するためのものではなく、
/// 端末を変えたときにデータを引き継ぐためだけのもの、と伝わる文言にしてある。
class SettingsScreen extends StatelessWidget {
  const SettingsScreen({super.key});

  Future<void> _signOut(BuildContext context, AuthController auth) async {
    final confirmed = await confirmAction(
      context,
      title: 'サインアウト',
      message: 'サインアウトしても記録は消えません。この端末から見えなくなるだけで、'
          '同じアカウントでサインインし直せば戻ります。',
      confirmLabel: 'サインアウト',
    );
    if (!confirmed) return;

    await auth.signOut();
  }

  Future<void> _reset(BuildContext context) async {
    final confirmed = await confirmAction(
      context,
      title: 'データを初期化',
      message: 'この端末のルーティンと記録をすべて削除して、初期状態に戻します。取り消せません。',
      confirmLabel: '初期化する',
    );
    if (!confirmed || !context.mounted) return;

    await context.read<AppState>().resetAll();
  }

  // 「同期」はもう無い。記録がサーバへ送れているかだけを見せる
  String _pendingLabel(AuthController auth) {
    if (auth.lastError != null) return '送信に失敗しました: ${auth.lastError}';
    if (auth.pendingCount > 0) return '未送信の記録: ${auth.pendingCount} 件';
    return 'すべて送信済みです';
  }

  @override
  Widget build(BuildContext context) {
    // Supabaseの初期化に失敗した場合や、認証を組み込まないテスト環境では
    // AuthController が存在しない。その場合もローカルモードとして通常どおり表示する。
    final auth = context.watch<AuthController?>();
    final mode = auth?.mode ?? StorageMode.local;
    final state = context.watch<AppState>();
    final counts = state.counts;

    return ListView(
      key: const Key('settings-screen'),
      padding: const EdgeInsets.all(Spacing.md),
      children: [
        const SectionTitle('データの保存先'),

        if (mode == StorageMode.local || auth == null)
          AppCard(
            key: const Key('settings-local-card'),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text('📱', style: TextStyle(fontSize: 28)),
                    const SizedBox(width: Spacing.md),
                    const Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(
                            'この端末専用の記録です',
                            style: TextStyle(
                              fontSize: 16,
                              fontWeight: FontWeight.bold,
                              color: AppColors.darkTextPrimary,
                            ),
                          ),
                          SizedBox(height: Spacing.xs),
                          Text(
                            'サインインしなくても全機能が使えます。ただしアプリを削除すると記録も消えます。',
                            style: TextStyle(
                              fontSize: 13,
                              height: 1.5,
                              color: AppColors.darkTextSecondary,
                            ),
                          ),
                        ],
                      ),
                    ),
                  ],
                ),
                const Divider(color: AppColors.darkBorder, height: Spacing.lg),
                const Text(
                  'メールアドレスを登録すると、機種変更やアプリの入れ直しをしても記録を引き継げます。'
                  '今ある記録はそのまま残ります（作り直しは起きません）。',
                  style: TextStyle(fontSize: 13, height: 1.5, color: AppColors.darkTextSecondary),
                ),
                const SizedBox(height: Spacing.md),
                AppButton(
                  key: const Key('settings-enable-backup'),
                  label: 'バックアップを有効にする',
                  onPressed: () => Navigator.of(context).push(
                    MaterialPageRoute(builder: (_) => const LoginScreen()),
                  ),
                ),
              ],
            ),
          )
        else
          AppCard(
            key: const Key('settings-cloud-card'),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text('☁️', style: TextStyle(fontSize: 28)),
                    const SizedBox(width: Spacing.md),
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          const Text(
                            'バックアップ有効',
                            style: TextStyle(
                              fontSize: 16,
                              fontWeight: FontWeight.bold,
                              color: AppColors.darkTextPrimary,
                            ),
                          ),
                          const SizedBox(height: Spacing.xs),
                          Text(
                            auth.viewer?.email ?? '',
                            key: const Key('settings-email'),
                            style: const TextStyle(
                              fontSize: 13,
                              color: AppColors.darkTextSecondary,
                            ),
                          ),
                        ],
                      ),
                    ),
                  ],
                ),
                const Divider(color: AppColors.darkBorder, height: Spacing.lg),
                Text(
                  _pendingLabel(auth),
                  key: const Key('settings-pending-status'),
                  style: TextStyle(
                    fontSize: 13,
                    color: auth.lastError != null
                        ? AppColors.darkError
                        : AppColors.darkTextSecondary,
                  ),
                ),
                const SizedBox(height: Spacing.md),
                AppButton(
                  key: const Key('settings-flush-now'),
                  label: '今すぐ送信',
                  variant: AppButtonVariant.secondary,
                  onPressed: auth.pendingCount == 0 ? null : auth.flushNow,
                ),
                const SizedBox(height: Spacing.sm),
                AppButton(
                  key: const Key('settings-sign-out'),
                  label: 'サインアウト',
                  variant: AppButtonVariant.ghost,
                  onPressed: () => _signOut(context, auth),
                ),
              ],
            ),
          ),

        const SizedBox(height: Spacing.lg),
        const SectionTitle('保存されているデータ'),
        AppCard(
          key: const Key('settings-counts'),
          child: Column(
            children: [
              _CountRow(label: 'ルーティン', value: counts.routines, valueKey: const Key('count-routines')),
              _CountRow(label: '種目', value: counts.exercises, valueKey: const Key('count-exercises')),
              _CountRow(label: 'ワークアウト記録', value: counts.routineLogs, valueKey: const Key('count-logs')),
              _CountRow(label: 'セット記録', value: counts.exerciseLogs, valueKey: const Key('count-sets')),
            ],
          ),
        ),

        const SizedBox(height: Spacing.lg),
        const SectionTitle('アプリについて'),
        const AppCard(
          child: _CountRow(label: 'バージョン', value: '1.0.0', valueKey: Key('settings-version')),
        ),

        const SizedBox(height: Spacing.lg),
        const SectionTitle('危険な操作'),
        AppCard(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                'この端末のデータをすべて消して初期状態に戻します。'
                'サーバ側も消えます。',
                style: const TextStyle(fontSize: 13, height: 1.5, color: AppColors.darkTextSecondary),
              ),
              const SizedBox(height: Spacing.md),
              AppButton(
                key: const Key('settings-reset'),
                label: 'データを初期化',
                variant: AppButtonVariant.danger,
                onPressed: () => _reset(context),
              ),
            ],
          ),
        ),
        const SizedBox(height: Spacing.xl),
      ],
    );
  }
}

class _CountRow extends StatelessWidget {
  final String label;
  final Object value;
  final Key valueKey;

  const _CountRow({required this.label, required this.value, required this.valueKey});

  @override
  Widget build(BuildContext context) => Padding(
        padding: const EdgeInsets.symmetric(vertical: Spacing.sm),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(label, style: const TextStyle(fontSize: 14, color: AppColors.darkTextSecondary)),
            Text(
              '$value',
              key: valueKey,
              style: const TextStyle(
                fontSize: 15,
                fontWeight: FontWeight.bold,
                color: AppColors.darkTextPrimary,
              ),
            ),
          ],
        ),
      );
}

File: nikkake_kmp/shared/src/commonMain/kotlin/com/myapplication/common/ui/screens/SettingsScreen.kt
package com.myapplication.common.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.Divider
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.myapplication.common.data.Collections
import com.myapplication.common.data.StorageMode
import com.myapplication.common.store.AppStore
import com.myapplication.common.store.AuthStore
import com.myapplication.common.ui.Destination
import com.myapplication.common.ui.Navigation
import com.myapplication.common.ui.components.AppButton
import com.myapplication.common.ui.components.AppCard
import com.myapplication.common.ui.components.ButtonVariant
import com.myapplication.common.ui.components.ConfirmDialog
import com.myapplication.common.ui.components.LabeledRow
import com.myapplication.common.ui.components.SectionTitle
import com.myapplication.common.ui.components.tag
import com.myapplication.common.ui.theme.LocalPalette
import com.myapplication.common.ui.theme.Spacing
import kotlinx.coroutines.launch

/**
 * 設定。
 *
 * この画面の主役は「バックアップするかどうか」の説明。
 * サインインは機能を開放するためのものではなく、
 * 端末を変えたときにデータを引き継ぐためだけのもの、と伝わる文言にしてある。
 */
@Composable
fun SettingsScreen(appStore: AppStore, authStore: AuthStore, navigation: Navigation) {
    val palette = LocalPalette.current
    val scope = rememberCoroutineScope()
    val counts = appStore.counts
    val mode = authStore.mode

    var confirmSignOut by remember { mutableStateOf(false) }
    var confirmReset by remember { mutableStateOf(false) }

    LazyColumn(
        modifier = Modifier.fillMaxSize().background(palette.background).tag("settings-screen"),
        contentPadding = PaddingValues(Spacing.md),
    ) {
        item {
            SectionTitle("データの保存先")

            if (mode == StorageMode.LOCAL) {
                AppCard(modifier = Modifier.tag("settings-local-card")) {
                    Column {
                        Row {
                            Text("📱", fontSize = 28.sp)
                            Spacer(Modifier.size(Spacing.md))
                            Column {
                                Text(
                                    "この端末専用の記録です",
                                    fontSize = 16.sp,
                                    fontWeight = FontWeight.Bold,
                                    color = palette.textPrimary,
                                )
                                Spacer(Modifier.height(Spacing.xs))
                                Text(
                                    "サインインしなくても全機能が使えます。ただしアプリを削除すると記録も消えます。",
                                    fontSize = 13.sp,
                                    color = palette.textSecondary,
                                )
                            }
                        }

                        Divider(color = palette.border, modifier = Modifier.padding(vertical = Spacing.md))

                        Text(
                            "メールアドレスを登録すると、機種変更やアプリの入れ直しをしても記録を引き継げます。" +
                                "今ある記録はそのまま残ります（作り直しは起きません）。",
                            fontSize = 13.sp,
                            color = palette.textSecondary,
                        )
                        Spacer(Modifier.height(Spacing.md))

                        AppButton(
                            label = "バックアップを有効にする",
                            onClick = { navigation.push(Destination.Login) },
                            modifier = Modifier.tag("settings-enable-backup"),
                        )
                    }
                }
            } else {
                AppCard(modifier = Modifier.tag("settings-cloud-card")) {
                    Column {
                        Row {
                            Text("☁️", fontSize = 28.sp)
                            Spacer(Modifier.size(Spacing.md))
                            Column {
                                Text(
                                    "バックアップ有効",
                                    fontSize = 16.sp,
                                    fontWeight = FontWeight.Bold,
                                    color = palette.textPrimary,
                                )
                                Spacer(Modifier.height(Spacing.xs))
                                Text(
                                    authStore.viewer?.email.orEmpty(),
                                    modifier = Modifier.tag("settings-email"),
                                    fontSize = 13.sp,
                                    color = palette.textSecondary,
                                )
                            }
                        }

                        Divider(color = palette.border, modifier = Modifier.padding(vertical = Spacing.md))

                        Text(
                            pendingLabel(authStore),
                            modifier = Modifier.tag("settings-pending-status"),
                            fontSize = 13.sp,
                            color = if (authStore.lastError != null) palette.error else palette.textSecondary,
                        )
                        Spacer(Modifier.height(Spacing.md))

                        AppButton(
                            label = "今すぐ送信",
                            variant = ButtonVariant.SECONDARY,
                            enabled = authStore.pendingCount > 0,
                            onClick = { scope.launch { authStore.flushNow() } },
                            modifier = Modifier.tag("settings-flush-now"),
                        )
                        Spacer(Modifier.height(Spacing.sm))
                        AppButton(
                            label = "サインアウト",
                            variant = ButtonVariant.GHOST,
                            onClick = { confirmSignOut = true },
                            modifier = Modifier.tag("settings-sign-out"),
                        )
                    }
                }
            }

            Spacer(Modifier.height(Spacing.lg))
            SectionTitle("保存されているデータ")
            AppCard(modifier = Modifier.tag("settings-counts")) {
                Column {
                    LabeledRow("ルーティン", "${counts.routines}", "count-routines")
                    LabeledRow("種目", "${counts.exercises}", "count-exercises")
                    LabeledRow("ワークアウト記録", "${counts.routineLogs}", "count-logs")
                    LabeledRow("セット記録", "${counts.exerciseLogs}", "count-sets")
                }
            }

            Spacer(Modifier.height(Spacing.lg))
            SectionTitle("アプリについて")
            AppCard {
                LabeledRow("バージョン", "1.0.0", "settings-version")
            }

            Spacer(Modifier.height(Spacing.lg))
            SectionTitle("危険な操作")
            AppCard {
                Column {
                    Text(
                        "ルーティンと記録をすべて削除して、初期状態に戻します。サーバ側も消えます。",
                        fontSize = 13.sp,
                        color = palette.textSecondary,
                    )
                    Spacer(Modifier.height(Spacing.md))
                    AppButton(
                        label = "データを初期化",
                        variant = ButtonVariant.DANGER,
                        onClick = { confirmReset = true },
                        modifier = Modifier.tag("settings-reset"),
                    )
                }
            }

            Spacer(Modifier.height(Spacing.xl))
        }
    }

    if (confirmSignOut) {
        ConfirmDialog(
            title = "サインアウト",
            message = "サインアウトしても記録は消えません。この端末から見えなくなるだけで、" +
                "同じアカウントでサインインし直せば戻ります。",
            confirmLabel = "サインアウト",
            onConfirm = {
                scope.launch { authStore.signOut() }
                confirmSignOut = false
            },
            onDismiss = { confirmSignOut = false },
        )
    }

    if (confirmReset) {
        ConfirmDialog(
            title = "データを初期化",
            message = "この端末のルーティンと記録をすべて削除して、初期状態に戻します。取り消せません。",
            confirmLabel = "初期化する",
            onConfirm = {
                scope.launch { appStore.resetAll() }
                confirmReset = false
            },
            onDismiss = { confirmReset = false },
        )
    }
}

// 「同期」はもう無い。記録がサーバへ送れているかだけを見せる
private fun pendingLabel(authStore: AuthStore): String = when {
    authStore.lastError != null -> "送信に失敗しました: ${authStore.lastError}"
    authStore.pendingCount > 0 -> "未送信の記録: ${authStore.pendingCount} 件"
    else -> "すべて送信済みです"
}
